src/util.js

```diff
@@ -16,4 +16,43 @@ export function processPotential(obj, resolve, reject){
       return reject(error);
     }
   }
-}
\ No newline at end of file
+}
+
+export function normalizeAbsolutePath(path, hasPushState) {
+    if (!hasPushState && path[0] !== '#') {
+      path = '#' + path;
+    }
+
+    return path;
+}
+
+export function createRootedPath(fragment, baseUrl, hasPushState) {
+  if (isAbsoluteUrl.test(fragment)) {
+    return fragment;
+  }
+
+  let path = '';
+
+  if (baseUrl.length && baseUrl[0] !== '/') {
+    path += '/';
+  }
+
+  path += baseUrl;
+
+  if (path[path.length - 1] != '/' && fragment[0] != '/') {
+    path += '/';
+  }
+
+  return normalizeAbsolutePath(path + fragment, hasPushState);
+}
+
+export function resolveUrl(fragment, baseUrl, hasPushState) {
+  if (isRootedPath.test(fragment)) {
+    return normalizeAbsolutePath(fragment, hasPushState);
+  } else {
+    return createRootedPath(fragment, baseUrl, hasPushState);
+  }
+}
+
+const isRootedPath = /^#?\//;
+const isAbsoluteUrl = /^([a-z][a-z0-9+\-.]*:)?\/\//i;
```

Comment:
This stuff just moved out of `router` for reuse in `navigation-plan`. The only changes were to parameterize the baseUrl and hasPushState instead of reading them off the current instance.